llvm/unittests/IR/ConstantFPRangeTest.cpp

```diff
@@ -0,0 +1,370 @@
+  EXPECT_EQ(PosZero.unionWith(NegZero), Zero);
+  EXPECT_EQ(NaN.unionWith(SNaN), NaN);
+  EXPECT_EQ(NaN.unionWith(QNaN), NaN);
+  EXPECT_EQ(SNaN.unionWith(QNaN), NaN);
+  EXPECT_EQ(
+      ConstantFPRange::getNonNaN(APFloat(1.0), APFloat(4.0))
+          .unionWith(ConstantFPRange::getNonNaN(APFloat(3.0), APFloat(6.0))),
+      ConstantFPRange::getNonNaN(APFloat(1.0), APFloat(6.0)));
+  EXPECT_EQ(
+      ConstantFPRange::getNonNaN(APFloat(1.0), APFloat(2.0))
+          .unionWith(ConstantFPRange::getNonNaN(APFloat(5.0), APFloat(6.0))),
+      ConstantFPRange::getNonNaN(APFloat(1.0), APFloat(6.0)));
+}
+
+TEST_F(ConstantFPRangeTest, FPClassify) {
+  EXPECT_EQ(Empty.classify(), fcNone);
+  EXPECT_EQ(Full.classify(), fcAllFlags);
+  EXPECT_EQ(Finite.classify(), fcFinite);
+  EXPECT_EQ(Zero.classify(), fcZero);
+  EXPECT_EQ(NaN.classify(), fcNan);
+  EXPECT_EQ(SNaN.classify(), fcSNan);
+  EXPECT_EQ(QNaN.classify(), fcQNan);
+  EXPECT_EQ(One.classify(), fcPosNormal);
+  EXPECT_EQ(Some.classify(), fcFinite);
+  EXPECT_EQ(SomePos.classify(), fcPosFinite);
+  EXPECT_EQ(SomeNeg.classify(), fcNegFinite);
+  EXPECT_EQ(PosInf.classify(), fcPosInf);
+  EXPECT_EQ(NegInf.classify(), fcNegInf);
+  EXPECT_TRUE(SomePos.toKnownFPClass().cannotBeOrderedLessThanZero());
+  EXPECT_EQ(Finite.getSignBit(), std::nullopt);
+  EXPECT_EQ(PosZero.getSignBit(), false);
+  EXPECT_EQ(NegZero.getSignBit(), true);
+  EXPECT_EQ(SomePos.getSignBit(), false);
+  EXPECT_EQ(SomeNeg.getSignBit(), true);
+  EXPECT_EQ(SomePos.toKnownFPClass().SignBit, false);
+  EXPECT_EQ(SomeNeg.toKnownFPClass().SignBit, true);
+
+  EnumerateConstantFPRanges([](const ConstantFPRange &CR) {
+    unsigned Mask = fcNone;
+    if (CR.containsSNaN())
+      Mask |= fcSNan;
+    if (CR.containsQNaN())
+      Mask |= fcQNan;
+    bool HasPos = CR.containsNaN(), HasNeg = CR.containsNaN();
+    APFloat Lower = CR.getLower();
+    const APFloat &Upper = CR.getUpper();
+    if (!(Lower.isPosInfinity() && Upper.isNegInfinity())) {
+      while (true) {
+        Mask |= Lower.classify();
+        HasPos |= !Lower.isNegative();
+        HasNeg |= Lower.isNegative();
+        if (Lower.bitwiseIsEqual(Upper))
+          break;
+        strictNext(Lower);
+      }
+    }
+
+    std::optional<bool> SignBit = std::nullopt;
+    if (HasPos != HasNeg)
+      SignBit = HasNeg;
+
+    EXPECT_EQ(SignBit, CR.getSignBit()) << CR;
+    EXPECT_EQ(Mask, CR.classify()) << CR;
+  });
+}
+
+TEST_F(ConstantFPRangeTest, Print) {
+  auto ToString = [](const ConstantFPRange &CR) {
+    std::string Str;
+    raw_string_ostream OS(Str);
+    CR.print(OS);
+    return Str;
+  };
+
+  EXPECT_EQ(ToString(Full), "full-set");
+  EXPECT_EQ(ToString(Empty), "empty-set");
+  EXPECT_EQ(ToString(NaN), "NaN");
+  EXPECT_EQ(ToString(SNaN), "SNaN");
+  EXPECT_EQ(ToString(QNaN), "QNaN");
+  EXPECT_EQ(ToString(One), "[1, 1]");
+  EXPECT_EQ(ToString(Some.unionWith(SNaN)), "[-3, 3] with SNaN");
+}
+
```

Comment:
Add a death test for the malformed ranges (e.g. lower bound greater than upper bound)?

Reply:
Done.